cql3/restrictions/statement_restrictions.cc

```diff
@@ -392,8 +392,9 @@ std::vector<const column_definition*> statement_restrictions::get_column_defs_fo
                 }
             }
         }
-        if (_clustering_columns_restrictions->needs_filtering(*_schema)) {
-            column_id first_filtering_id = _schema->clustering_key_columns().begin()->id +
+        const bool pk_has_unrestricted_components = _partition_key_restrictions->has_unrestricted_components(*_schema);
+        if (pk_has_unrestricted_components || _clustering_columns_restrictions->needs_filtering(*_schema)) {
+            column_id first_filtering_id = pk_has_unrestricted_components ? 0 : _schema->clustering_key_columns().begin()->id +
```

Comment:
The new code looks correct.
I think it demonstrates that clustering_column_restrictions->needs_filtering() is a bad API, because the needs_filtering() function needs both the pk restrictions and the ck restrictions to make the right decision. So we should have had a free function taking both pk and ck restrictions. But I'm not insisting that you do that, your fix is also fine.

Reply:
I agree, it's simply yet another reason for rewriting restrictions to be not object-oriented - clustering restrictions themselves simply do not know if they can use filtering/indexing, because it depends on other existing restrictions. For now I'm leaving this code as is, because this bug is higher priority than big code rewrite.